File: js/packages/tur-edgy/src/tween.ts
import type { Color } from "./index";

// `__tur.colorLerp` is registered alongside `createColor` in the Rust bridge
// (`core/bridge/color.rs::tur_color_lerp`) and delegates to `Color::lerp` in
// tur-shared. We declare only the slice of `__tur` this module needs.
declare const __tur: {
    colorLerp(a: Color, b: Color, t: number): Color;
};

/**
 * A linear interpolation between a beginning and ending numeric value,
 * mirroring Flutter's `Tween<double>`.
 *
 * Used with an `AnimationController`'s `onTick` to drive a source atom:
 *
 * ```ts
 * const width$ = source(100);
 * const tween = Tween({ begin: 100, end: 280 });
 * const ctrl = createAnimationController({
 *     duration: 1000,
 *     onTick: mutate((_ctx, t) => set(width$, tween.lerp(t))),
 * });
 * ctrl.forward();
 * ```
 *
 * `begin` and `end` are mutable (getters + setters), matching Flutter's
 * Tween — changes are honoured the next time `lerp` runs.
 */
export interface TweenValue {
    /** The value at the start of the animation (`t = 0`). Mutable. */
    begin: number;
    /** The value at the end of the animation (`t = 1`). Mutable. */
    end: number;
    /** Interpolate at parameter `t`. `t` is NOT clamped (matches Flutter). */
    lerp(t: number): number;
    /** Interpolate at parameter `t`, clamped to `[0, 1]`. */
    transform(t: number): number;
}

export function Tween(opts: { begin: number; end: number }): TweenValue {
    let begin = opts.begin;
    let end = opts.end;
    return {
        get begin() {
            return begin;
        },
        set begin(v: number) {
            begin = v;
        },
        get end() {
            return end;
        },
        set end(v: number) {
            end = v;
        },
        lerp(t: number): number {
            return begin + (end - begin) * t;
        },
        transform(t: number): number {
            return begin + (end - begin) * Math.max(0, Math.min(1, t));
        },
    };
}

/**
 * A linear interpolation between two `Color` values, mirroring Flutter's
 * `ColorTween` (which delegates to `Color.lerp`). Channels are interpolated
 * component-wise in u8 space on the Rust side via `__tur.colorLerp`.
 *
 * `begin` and `end` are mutable, matching Flutter.
 */
export interface ColorTweenValue {
    begin: Color;
    end: Color;
    lerp(t: number): Color;
    transform(t: number): Color;
}

export function ColorTween(opts: {
    begin: Color;
    end: Color;
}): ColorTweenValue {
    let begin = opts.begin;
    let end = opts.end;
    return {
        get begin() {
            return begin;
        },
        set begin(v: Color) {
            begin = v;
        },
        get end() {
            return end;
        },
        set end(v: Color) {
            end = v;
        },
        lerp(t: number): Color {
            return __tur.colorLerp(begin, end, t);
        },
        transform(t: number): Color {
            return __tur.colorLerp(begin, end, Math.max(0, Math.min(1, t)));
        },
    };
}

File: libs/tur-animation/src/tween.rs
//! `Tween<T>` — linear interpolation between a beginning and ending value,
//! mirroring Flutter's [`Tween<T>`] abstraction.
//!
//! A `Tween` defines a value range `[begin, end]` and produces intermediate
//! values via [`Tween::lerp`] (and its alias [`Tween::transform`]) given a
//! normalized parameter `t` in `[0, 1]`. Unlike [`crate::Curve`] (which
//! reshapes *time*), a `Tween` interpolates *values*.
//!
//! The generic [`NumTween`] handles numeric types whose default `lerp` is
//! `begin + (end - begin) * t`. Types needing specialized interpolation ship
//! dedicated tweens (e.g. [`ColorTween`], which interpolates RGBA channels
//! component-wise like Flutter's `ColorTween`/`Color.lerp`).
//!
//! [`Tween<T>`]: https://api.flutter.dev/flutter/animation/Tween-class.html

use tur_engine::core::render::Color;

/// A linear interpolation between a beginning and ending value of type `T`.
///
/// Implementors define [`Tween::lerp`]; [`Tween::transform`] has a default
/// impl that clamps `t` to `[0, 1]` then delegates to `lerp` (matching
/// Flutter's `Tween.transform`, which clamps negative time but passes values
/// ≥ 1 through — here we clamp both ends for predictability in layout code).
pub trait Tween<T: Clone>: Clone {
    /// The value at parameter `t` (typically an eased progress in `[0, 1]`).
    fn lerp(&self, t: f64) -> T;

    /// Default-implemented in terms of [`lerp`](Tween::lerp), with `t` clamped
    /// to `[0, 1]`. Override only if a tween needs extrapolation semantics.
    fn transform(&self, t: f64) -> T {
        self.lerp(t.clamp(0.0, 1.0))
    }

    /// Convenience: evaluate at the current value of an `AnimationController`-
    /// style progress. Equivalent to Flutter's `Animatable::evaluate`.
    fn evaluate(&self, t: f64) -> T {
        self.transform(t)
    }
}

/// Generic numeric tween. The default Flutter `Tween<T>` for types satisfying
/// the `+`, `-`, `* double` contract — here specialized to `f64` since that's
/// the universal numeric prop type in tur.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct NumTween {
    begin: f64,
    end: f64,
}

impl NumTween {
    pub const fn new(begin: f64, end: f64) -> Self {
        Self { begin, end }
    }
}

impl Tween<f64> for NumTween {
    fn lerp(&self, t: f64) -> f64 {
        self.begin + (self.end - self.begin) * t
    }
}

/// Tween between two [`Color`] values, interpolating RGBA channels
/// component-wise in u8 space (matching Flutter's `Color.lerp` /
/// `ColorTween`). Like Flutter, if either side is unspecified the other side
/// is returned unchanged at the appropriate endpoint; here both `begin` and
/// `end` are non-nullable (tur's `Color` is opaque).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ColorTween {
    begin: Color,
    end: Color,
}

impl ColorTween {
    pub const fn new(begin: Color, end: Color) -> Self {
        Self { begin, end }
    }
}

impl Tween<Color> for ColorTween {
    fn lerp(&self, t: f64) -> Color {
        Color::lerp(self.begin, self.end, t)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn num_tween_endpoints_and_midpoint() {
        let tw = NumTween::new(100.0, 200.0);
        assert_eq!(tw.lerp(0.0), 100.0);
        assert_eq!(tw.lerp(1.0), 200.0);
        assert_eq!(tw.lerp(0.5), 150.0);
        assert_eq!(tw.lerp(0.25), 125.0);
    }

    #[test]
    fn num_tween_negative_range() {
        let tw = NumTween::new(200.0, 100.0);
        assert_eq!(tw.lerp(0.0), 200.0);
        assert_eq!(tw.lerp(1.0), 100.0);
        assert_eq!(tw.lerp(0.5), 150.0);
    }

    #[test]
    fn transform_clamps_out_of_range_t() {
        let tw = NumTween::new(0.0, 10.0);
        assert_eq!(tw.transform(-1.0), 0.0);
        assert_eq!(tw.transform(2.0), 10.0);
    }

    #[test]
    fn color_tween_endpoints_exact() {
        let a = Color::rgb(0, 0, 0);
        let b = Color::rgb(255, 255, 255);
        let tw = ColorTween::new(a, b);
        assert_eq!(tw.lerp(0.0), a);
        assert_eq!(tw.lerp(1.0), b);
    }

    #[test]
    fn color_tween_midpoint_rounds() {
        let a = Color::rgb(0, 0, 0);
        let b = Color::rgb(100, 200, 50);
        let mid = ColorTween::new(a, b).lerp(0.5);
        assert_eq!(mid.r(), 50);
        assert_eq!(mid.g(), 100);
        assert_eq!(mid.b(), 25);
    }

    #[test]
    fn color_tween_preserves_alpha_channel() {
        let a = Color::rgba(10, 20, 30, 40);
        let b = Color::rgba(110, 220, 130, 240);
        let mid = ColorTween::new(a, b).lerp(0.5);
        assert_eq!(mid.r(), 60);
        assert_eq!(mid.g(), 120);
        assert_eq!(mid.b(), 80);
        assert_eq!(mid.a(), 140);
    }
}
